src/change_stream/mod.rs

```diff
@@ -236,50 +240,56 @@ where
+                        return Poll::Pending;
+                    }
+                    Poll::Ready(Ok(new_stream)) => {
+                        // Ensure that the old cursor is killed on the server selected for the new
+                        // one.
+                        self.cursor
+                            .set_drop_address(new_stream.cursor.address().clone());
+                        self.cursor = new_stream.cursor;
+                        self.args = new_stream.args;
+                        continue;
+                    }
+                    Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
                 }
-                Poll::Ready(Ok(new_stream)) => {
-                    // Ensure that the old cursor is killed on the server selected for the new one.
-                    self.cursor
-                        .set_drop_address(new_stream.cursor.address().clone());
-                    self.cursor = new_stream.cursor;
-                    self.args = new_stream.args;
-                    return Poll::Pending;
-                }
-                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
             }
-        }
-        let out = self.cursor.poll_next_in_batch(cx);
-        match &out {
-            Poll::Ready(Ok(bv)) => {
-                if let Some(token) = get_resume_token(bv, self.cursor.post_batch_resume_token())? {
-                    self.data.resume_token = Some(token);
+            let out = self.cursor.poll_next_in_batch(cx);
+            match &out {
+                Poll::Ready(Ok(bv)) => {
+                    if let Some(token) =
+                        get_resume_token(bv, self.cursor.post_batch_resume_token())?
+                    {
+                        self.data.resume_token = Some(token);
+                    }
+                    if matches!(bv, BatchValue::Some { .. }) {
+                        self.data.document_returned = true;
+                    }
                 }
-                if matches!(bv, BatchValue::Some { .. }) {
-                    self.data.document_returned = true;
+                Poll::Ready(Err(e)) if e.is_resumable() && !self.data.resume_attempted => {
+                    self.data.resume_attempted = true;
+                    let client = self.cursor.client().clone();
+                    let args = self.args.clone();
+                    let mut data = self.data.take();
+                    data.implicit_session = self.cursor.take_implicit_session();
+                    self.pending_resume = Some(Box::pin(async move {
+                        let new_stream: Result<ChangeStream<ChangeStreamEvent<()>>> = client
+                            .execute_watch(args.pipeline, args.options, args.target, Some(data))
+                            .await;
+                        new_stream.map(|cs| cs.with_type::<T>())
+                    }));
+                    // Iterate the loop so the new future gets polled and can register wakers.
```

Comment:
This was the point of wrapping the logic in a loop - there needs to be an underlying future that's been polled rather than just returning `Poll::Pending` directly because with the latter this future will never be polled again due to the lack of registered wake conditions.